collections/src/main/scala/strawman/collection/immutable/LazyList.scala

```diff
@@ -246,28 +254,13 @@ sealed private[immutable] trait LazyListOps[+A, +CC[+X] <: LinearSeq[X] with Laz
     * @return The lazy list containing elements of this lazy list and the iterable object.
     */
   def lazyAppendAll[B >: A](suffix: => collection.IterableOnce[B]): CC[B] =
-    if (isEmpty) iterableFactory.from(suffix.iterator()) else cons[B](head, tail.lazyAppendAll(suffix))
+    if (isEmpty) iterableFactory.from(suffix) else cons[B](head, tail.lazyAppendAll(suffix))
```

Comment:
Actually if `suffix` is already a `LazyList`, then `iterableFactory.from(suffix)` is a no-op, so I would keep using just `suffix`.

Reply:
It seems it fixes #421.

Reply:
Interesting, do you have any insight of why? Looking at the code this is not clear to me.

Reply:
With `suffix.iterator()` stacktrace is (differs from original issue)
```
java.lang.StackOverflowError
  at strawman.collection.immutable.LazyList$Cons.head$lzycompute(LazyList.scala:462)
  at strawman.collection.immutable.LazyList$Cons.head(LazyList.scala:460)
  at strawman.collection.LinearSeqOps$$anon$1.next(LinearSeq.scala:25)
  at strawman.collection.Iterator$$anon$14.next(Iterator.scala:891)
  at strawman.collection.immutable.LazyList$.evaluatedElem$lzycompute$1(LazyList.scala:538)
  at strawman.collection.immutable.LazyList$.evaluatedElem$1(LazyList.scala:538)
  at strawman.collection.immutable.LazyList$.$anonfun$fromIterator$1(LazyList.scala:539)
  at strawman.collection.immutable.LazyList$Cons.head$lzycompute(LazyList.scala:462)
  at strawman.collection.immutable.LazyList$Cons.head(LazyList.scala:460)
  at strawman.collection.immutable.LazyListOps.flatMap(LazyList.scala:340)
  at strawman.collection.immutable.LazyListOps.flatMap$(LazyList.scala:335)
  at strawman.collection.immutable.LazyList.flatMap(LazyList.scala:186)
  at strawman.collection.immutable.LazyListOps.$anonfun$flatMap$1(LazyList.scala:348)
  at strawman.collection.immutable.LazyListOps.lazyAppendAll(LazyList.scala:257)
  at strawman.collection.immutable.LazyListOps.lazyAppendAll$(LazyList.scala:256)
  at strawman.collection.immutable.LazyList.lazyAppendAll(LazyList.scala:186)
  at strawman.collection.immutable.LazyListOps.$anonfun$lazyAppendAll$2(LazyList.scala:257)
  at strawman.collection.immutable.LazyList$Cons.tail$lzycompute(LazyList.scala:466)
  at strawman.collection.immutable.LazyList$Cons.tail(LazyList.scala:464)
  at strawman.collection.immutable.LazyList$Cons.tail(LazyList.scala:456)
  at strawman.collection.LinearSeqOps$$anon$1.next(LinearSeq.scala:25)
  at strawman.collection.immutable.LazyList$.evaluatedElem$lzycompute$1(LazyList.scala:538)
  at strawman.collection.immutable.LazyList$.evaluatedElem$1(LazyList.scala:538)
  at strawman.collection.immutable.LazyList$.$anonfun$fromIterator$1(LazyList.scala:539)
  at strawman.collection.immutable.LazyList$Cons.head$lzycompute(LazyList.scala:462)
  at strawman.collection.immutable.LazyList$Cons.head(LazyList.scala:460)
  at strawman.collection.LinearSeqOps$$anon$1.next(LinearSeq.scala:25)
  at strawman.collection.immutable.LazyList$.evaluatedElem$lzycompute$1(LazyList.scala:538)
  at strawman.collection.immutable.LazyList$.evaluatedElem$1(LazyList.scala:538)
  at strawman.collection.immutable.LazyList$.$anonfun$fromIterator$1(LazyList.scala:539)
  at strawman.collection.immutable.LazyList$Cons.head$lzycompute(LazyList.scala:462)
  at strawman.collection.immutable.LazyList$Cons.head(LazyList.scala:460)
```
If I understood it correctly then methods are called in such order (unused lines are hid under comments):
```scala
def flatMap[B](f: A => IterableOnce[B]): CC[B] =
  // if (isEmpty) iterableFactory.empty
  // else 
  {
    var nonEmptyPrefix: CC[A] = coll
    var prefix = iterableFactory.from(f(nonEmptyPrefix.head))
    while (!nonEmptyPrefix.isEmpty && prefix.isEmpty) {
      // nonEmptyPrefix = nonEmptyPrefix.tail
      // if(!nonEmptyPrefix.isEmpty)
      //   prefix = iterableFactory.from(f(nonEmptyPrefix.head))
    }

    // if (nonEmptyPrefix.isEmpty) iterableFactory.empty
    // else 
    prefix.lazyAppendAll(nonEmptyPrefix.tail.flatMap(f))
  }

def lazyAppendAll[B >: A](suffix: => collection.IterableOnce[B]): CC[B] =
    if (isEmpty) iterableFactory.from(suffix.iterator()) else cons[B](head, tail.lazyAppendAll(suffix))

def fromIterator[A](it: Iterator[A]): LazyList[A] =
  // if (it.hasNext) {
    lazy val evaluatedElem = it.next()
    new LazyList.Cons(evaluatedElem, { evaluatedElem; fromIterator(it) })
  // } else LazyList.Empty
```
`fromIterator` is recursive (without 'tail-') and it's a cause of SOE.

Reply:
I see, then we should fix these operations to be stack-safe. But I think that should be done in a separate PR. In the meantime, I continue thinking that it’s better to use `suffix` rather than `suffix.iterator()`.

Reply:
I think `.iterator()`  is redundant in any case. That's the reason I removed it.

Should I open issue for `fromIterator`?

Reply:
> I think .iterator() is redundant in any case. That's the reason I removed it.

OK, actually I don’t know why I read the diff in the opposite side: I thought you had added the `.iterator` calls… But I was wrong, your PR is correct :)

> Should I open issue for fromIterator?

Sure!